atlasdb-commons/src/main/java/com/palantir/common/concurrent/ConcurrentMaps.java

```diff
@@ -34,7 +39,18 @@ public final class ConcurrentMaps {
      * @param expectedEntries number of expected entries
      */
     public static <K, V> ConcurrentMap<K, V> newWithExpectedEntries(int expectedEntries) {
-        return new ConcurrentHashMap<>(expectedEntries * (/* contention probability */ 8 * /* load factor */ 3 / 4));
+        int concurrencyLevel = Math.max(1, Runtime.getRuntime().availableProcessors());
+        return new ConcurrentHashMap<>(initialCapacity(expectedEntries), DEFAULT_LOAD_FACTOR, concurrencyLevel);
+    }
+
+    @VisibleForTesting
+    static int initialCapacity(int expectedEntries) {
+        return Math.max(0, (int) ((expectedEntries * (INVERSE_CONTENTION_PROBABILITY * DEFAULT_LOAD_FACTOR)) - 1));
```

Comment:
That said, I'm still not 100% sure I understand why we chose `INVERSE_CONTENTION_PROBABILTY * expectedEntries` as the number of buckets. 

I think I see the logic in the CHM discussion, but because we're so far below the usual load factor (we'd be at 0.125), I'd expect the number of elements per bin under random hashing to then follow a Poisson distribution with parameter 0.125 (not 0.5 as mentioned in their calculations, which if I'm reading correctly is based on a load factor of 0.75 and resizing). My read there is that the "lock contention probability is 1 / 8 * number of elements" thing relates to _that_ parameter 0.5 distribution. 

Setting ourselves to use 8 times the expected number of elements in terms of buckets should indeed mean that we rarely contend, but I don't know if this was very specifically intentional (in which case I'm not fully sure I understand the intention), or meant as a general heuristic. My read on this is that the latter, but clarity here would be appreciated.

I think given in the worst case this uses 2048 * a few bytes, the impact on memory is in the order of kilobytes, and so this is worth a shot anyway though.